src/zope/testrunner/runner.py

```diff
@@ -898,13 +898,30 @@ def testTearDown(self):
             if hasattr(layer, 'testTearDown'):
                 layer.testTearDown()
 
+    def _makeBufferedStdStream(self):
+        """Make a buffered stream to replace a standard stream."""
+        if sys.version_info[0] >= 3:
+            # The returned stream needs to have a 'buffer' attribute, since
+            # some tests may expect that to exist on standard streams, and a
+            # 'getvalue' method for the convenience of _restoreStdStreams.
+            # This requires some care.
+            class BufferedStandardStream(io.TextIOWrapper):
+                def getvalue(self):
+                    return self.buffer.getvalue().decode(
+                        encoding=self.encoding, errors=self.errors)
+
+            return BufferedStandardStream(
+                io.BytesIO(), newline='\n', write_through=True)
+        else:
+            return StringIO()
```

Comment:
`StringIO` is `from six import StringIO`. On Python 2, that's `StringIO.StringIO`, which is the non-native version (`cStringIO.StringIO` is the native version). 

Long story short, I was concerned about possible performance implications, but then I saw the next lines where we were unconditionally assigning `StringIO` instances, so this shouldn't be a regression. Indeed, on Python 3 it might be an improvement!

Reply:
Yes, it should be near enough identical on Python 2, and on Python 3 I think `io.StringIO` and an `io.TextIOWrapper` around an `io.BytesIO` are likely to perform similarly enough for our purposes.

I've taken to explicitly using `six.StringIO` or `io.StringIO` in new code rather than doing something like `from six import StringIO`, because the fact that there are multiple kinds of `StringIO` with different semantics is far too confusing (in hindsight it's a shame that the modern version wasn't called `io.TextIO`).  Cleaning that up can be for later though.